src/recovery/recover_entity.rs

```diff
@@ -0,0 +1,1021 @@
+use std::{hash::Hash, ops::Add, sync::RwLock};
+
+use crate::prelude::*;
+
+impl Profile {
+    async fn new_entity<E: IsEntity + std::fmt::Debug + std::hash::Hash + Eq>(
+        &mut self,
+        network_id: NetworkID,
+        name: impl AsRef<str>,
+        factor_source_id: FactorSourceIDFromHash,
+        gateway: Arc<dyn Gateway>,
+    ) -> E {
+        assert!(self
+            .factor_sources
+            .iter()
+            .map(|f| f.factor_source_id())
+            .contains(&factor_source_id));
+
+        let entity_kind = E::kind();
+        let key_kind = CAP26KeyKind::T9n;
+        let key_space = KeySpace::Unsecurified;
+        let name = name.as_ref();
+
+        let index_assigner = NextFreeIndexAssigner::live();
+
+        let base =
+            index_assigner.derivation_index_for_factor_source(NextFreeIndexAssignerRequest {
+                network_id,
+                factor_source_id,
+                key_space,
+                entity_kind,
+                profile: self,
+            });
+
+        let mut genesis_factor_and_address: Option<(
+            HierarchicalDeterministicFactorInstance,
+            E::Address,
+        )> = None;
+        for index in base..(base.add_n(50)) {
```

Comment:
this whole loop should go away, should just use the next factor from the `pre_derive_keys_cache` (not implemented)